### ISSUE TITLE
Create a resource for developers new to rust networking

### ISSUE THREAD

**Issue by Yarn** [rustasync/team#37]
Currently much of the ecosystem is in flux and it can be difficult to determine what crates are suitable for production, are stable, being phased out, work together, etc.

This would likely be part of the working group website (https://github.com/rust-lang-nursery/wg-net/issues/21)

There are some resources that work toward this goal already.
https://github.com/rust-unofficial/awesome-rust#web-programming
http://www.arewewebyet.org/
https://github.com/flosse/rust-web-framework-comparison
https://wiki.alopex.li/AnOpinionatedGuideToRustWebServers

I think http://www.arewewebyet.org/ is the closest to what we'd want to make but covering networking more broadly and more opinionated/updated (e.g. the first crate listed for json is rustc-serialize despite it being deprecated for over a year)
Ideally we can give some example stacks for specific use cases and say we know these work together and you can put them in production (and provide references to any documentation needed).

Edit: https://wiki.alopex.li/AnOpinionatedGuideToRustWebServers is the type of opinionated information about a framework I'd like to see. We would probably want to have feedback that reflects some consensus and not just have one person trying a bunch of frameworks.

---

Beyond just listing what frameworks are available there is a lot of ecosystem level information that isn't immediately obvious that would be nice to consolidate into a reference for people who aren't familiar with the situation. This includes things like ring/openssl dependency hell and what versions of libs will work together or what versions/crates to use if you are using sync or async (especially if more things do what hyper did and have sync/async versions that aren't mentioned anywhere obvious).

A general overview of any ongoing transitions in the ecosystem is probably also valuable. This could just be larger things like moving to async in general but could also cover adoption of new versions of widely used crates like ring 0.11 -> 0.13 or futures 0.1 -> 0.3 and what versions work for commonly used combinations of crates. (tracking adoption of new versions of dependencies across the ecosystem is obviously a non-trivial task but currently the task is just shifted to individual developers in a lot of cases)

---

There has also been discussion of writing guides for getting started with various networking topics. For guides on specific crates I think we should focus on making the crate's docs more discoverable and/or contributing to them directly but including guides in cases where the topic doesn't clearly belong in a given crates docs makes sense.
meta: should guides and general ecosystem information be separate issues?

---

This is obviously strongly related to tracking what crates are in the ecosystem and the general state of the ecosystem which has already been started in other tickets.
https://github.com/rust-lang-nursery/wg-net/issues/27
https://github.com/rust-lang-nursery/wg-net/issues/28

**Comment by Yarn** [comment]
shameless repost of https://github.com/rust-lang-nursery/wg-net/issues/35#issuecomment-408969417 by @bIgBV 

One thing I have some thoughts on is the general approach to introducing new programmers to networking in rust. Let me know if this isn't the right forum for this, I'll create a new issue/post on a different one if required. Sync or async. Right now, as far as I can tell there isn't one place a programmer new to rust can go to get a high-level view of the ecosystem. There is great documentation in getting started with rust, and individual projects have great documentation, but unless you know of them already, you wouldn't know how to find them.

## Proposal
* I propose that we have a website for rust networking/a page on the official language website. * The topics should be split up based on which layer of the network stack they pertain to.
* The information presented by these topics could be as such:
    * Information regarding getting started with building something at that layer 
    * Point out the current best practices and point to the libraries recommended by the community
    * Point to further resources that someone can use when working at that layer

I feel like this approach will cater to users at all levels of the networking story. When the website is first built, the topics needn't have extensive examples and documentation but just point to the right resources. This would mean that someone who wants to build a web application in rust wouldn't have to search for Reddit threads titled "what library do you use to build web applications", which I am guessing a lot of people currently do, and I myself have done at one point.

I used web applications just as an example, the same idea could be applied to people writing protocols at the application layer or even lower when writing custom protocols on top of IP.

## Conclusion
Having a central website for everyone to start their networking journey at would be great at helping programmers find the information they are looking for and also help the community consolidate all the information regarding networking in rust.

**Comment by ALCC01** [comment]
I would support having a arewe(.+)yet style website to allow newcomers to quickly grasp the state of the ecosystem, providing an extract of the relevant guidelines (such as stacks, recommended crates, ...) and linking to the documentation/best practices/recommendations for each topic. These topics may fall under the responsibility of the wg's subgroups, so that we can assure some degree of accuracy and coherence.

This website may serve as a first point of contact to further explore relevant documentation (I think slapping a book/documentation in the user's face might not be most approachable method, it may be better providing a TLDR and then link into the documentation).

**Comment by gruberb** [comment]
I like the idea of giving newcomers a place to start. However, keeping this website up-to-date seems tedious. Rust is quite young and very dynamic.

Why not go a step further and explain each layer in a "technical" detail (in Mozilla code comics style maybe) and explain what needs to be done in each layer to support a full fledged server application.

We can then give recommendations of how to tag rust repositories on GitHub so the creators of solutions can update and tag their own solutions. So we keep the updated part on them but provide a much more detailed few on things.

**Pro:**
- it would be even a general information/learning ressource to understand server applications etc.
- People will get drawn to this documentation and discover Rust along the way
- The documentation has a much longer life cycle

**Contra:**
- Much technical knowledge is needed as well as be able to explain it in a newcomer friendly way. 

As for me, I come from NodeJS and Go, and I was happy to see and read the discussions around server side programming in Rust, because it gave me a reason to dig deeper into the topic and understand it on a deeper level.

**Comment by Yarn** [comment]
>However, keeping this website up-to-date seems tedious. Rust is quite young and very dynamic.

The thing is, while it might be tedious for us, it's a major roadblock to getting started for someone coming from outside the ecosystem. Between C dependencies, sync vs async, different futures versions etc. it can be a major challenge to even get all the different things you need working together.

On top of that most people want to know what framework they'd actually want to use. Different libraries will be better for different use cases (or sometimes one will just have better ergonomics/performance/interop)and currently this is just cultural knowledge. If someone wants to use rust for work they are unlikely to engage the community extensively to find out what the consensus on frameworks is or to try a bunch of them in enough depth to choose one. If they choose wrong the experience will be bad and the idea that rust isn't ergonomic or is dependency hell will be engrained in their workplace or social group.

For example google "rust web framework" and "python web framework". Python has for more resources actually providing some reasonable comparison between frameworks. Beyond this rust has a result about it being hard to figure out which framework to use (https://news.ycombinator.com/item?id=16694456). arewewebyet shows up which isn't bad and at least only gives you 5 options to choose randomly between which just leaves trying to find a password hashing crate that doesn't have a dependency conflict with whatever you randomly chose (but you may not know to do that last part until you try enabling TLS, better hope you don't have to switch to an older major version of your framework).
(this is just for web frameworks, I would imagine things don't get better lower in the stack except maybe having less than 5 choices to begin with)

Tagging rust repos would make getting a list of crates for a given thing easier but you still have no idea which ones are good/maintained and doesn't solve any issues with incompatible libs.

That turned into a bit of a rant.

**tl;dr** We need to maintain a resource like this _because_ it's tedious. Not having this resource just shifts the work to individuals (unless they just choose not to use rust).

To respond to every other sentence in your post besides the second and third, I think resources teaching low level networking with rust are a great idea (largely because they are something I want personally and am figuring at least some people will share the sentiment). I suspect there is some non trivial userbase that would be interested in learning lower level networking, systems programming, embedded development if they didn't have to spend most of their time trying to figure out segfaults and indecipherable C++ errors.

**Comment by bIgBV** [comment]
@Yarn I completely agree with that. Sure the maintenance of such a website would be tedious, but at least it would give us something to point to when someone asks how to get started with web programming, instead of it being a different answer based on who you ask.

I also like the idea of having resources for teaching low level networking in rust. The topics on the website could be split according to the different networking layers. But I suppose the areas where most people would be interested would be the application layer and the transport layer.
